### PR TITLE
cleaned compiler warnings; ported ADD instruction

### PR DESCRIPTION
**Pull request by williamberman**

### CODE REVIEW

File: core/src/memory/mod.rs [hunk 1]
```diff
@@ -3,23 +3,20 @@ pub mod memory_models;
...
 use error::MemoryError;
-use z3::{ast::Array, FuncDecl};
 
-pub type MemoryResult<T> = Result<T, MemoryError>;
-pub trait ReadOnlyMem: Sized {
+pub trait Mem: Sized {
     type MemVal;
     type Index;
-    fn read(&self, idx: Self::Index) -> MemoryResult<Option<Self::MemVal>>;
 }
```
Comment (williamberman) [inline core/src/memory/mod.rs:13]: Added Mem trait. Useful like in the case of ADD where memory isn't touched

File: contrib/src/instructions/arith.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            changed: vec![
+                StackOpRecord::Pop(op_1),
+                StackOpRecord::Pop(op_2),
+                StackOpRecord::Push(res),
+            ],
+        });
+
+        Ok(change_log)
+    }
+}
```
Comment (williamberman) [inline contrib/src/instructions/arith.rs:41]: ADD instruction ported over using `AbstractInstruction` trait

File: core/src/memory/mod.rs (hunk 1 above)
Comment (williamberman) [reply core/src/memory/mod.rs:13]: Also allows machine/instruction trait to be less restrictive (not require RWMem)

File: contrib/src/instructions/arith.rs (hunk 2 above)
Comment (WilfredTA) [reply contrib/src/instructions/arith.rs:41]: I actually am becoming more convinced that implementing the trait for each Instruction as opposed to for an enum of all instructions is the right approach, as it allows us to be express more heterogeneous behaviors within a single instruction set with less type/trait gymnastics in the core library.

But what *is* interesting is that we could really do it either way it seems

File: core/tests/simple_lang.rs [hunk 3]
```diff
@@ -94,7 +94,7 @@ impl<'a> VMInstruction<'a> for Instruction<Int<'a>> {
                 let prev_val = {
                     match memory.read(mem_offset.clone()) {
                         Ok(val) => val.unwrap(),
-                        Err(e) => Int::from_u64(val.get_ctx(), 0),
+                        Err(_e) => Int::from_u64(val.get_ctx(), 0),
```
Comment (WilfredTA) [inline core/tests/simple_lang.rs:97]: nit: can just do `Err(_)` here instead

File: contrib/src/instructions/arith.rs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        stack: &S,
+        _mem: &M,
+        _ext: &Extension,
+    ) -> InstructionResult<AbstractExecRecord<S, M, <Extension as EnvExtension>::DiffRecordType, C>>
+    where
+        C: Into<symbolic_stack_machines_core::constraint::Constraint<C>>,
+    {
+        let mut change_log = AbstractExecRecord::default();
+
+        let op_1: T = stack.peek(0).unwrap();
```
Comment (WilfredTA) [inline contrib/src/instructions/arith.rs:27]: Wondering actually how we should handle the case wherein `stack.peek(x)` is `None`. Perhaps this should result in adding a constraint that the values do exist on the stack... Perhaps this already happens when the diff is processed by the machine (so, this is a note-to-self to look into that).

**Review by WilfredTA — APPROVED** [review]

File: contrib/src/instructions/arith.rs (hunk 4 above)
Comment (williamberman) [reply contrib/src/instructions/arith.rs:27]: I assume there's a standard stack underflow error. I would just throw that for now and assume all initial implementations will have concrete stack size

File: contrib/src/instructions/arith.rs (hunk 2 above)
Comment (williamberman) [reply contrib/src/instructions/arith.rs:41]: Yep +1 because there's just a trait to implement we can do either way

**Pull request merged**